docs/proposals/29-autoscaling-storage-volumes.md

```diff
@@ -0,0 +1,335 @@
+default settings, was:
+- Prometheus average utilisation: 16.6%; &plusmn;1&sigma;: (8.0 - 25.2)
+- Vali average utilisation: 8.3%; &plusmn;1&sigma;: (4.2 - 12.3)
+
+The above aggregations include both shoot and seed volumes.
+
+### Limitations
+#### Reliance on Prometheus
+`pvc-autoscaler` relies on metrics from the seed's cache Prometheus instance. The availability of the otherwise simple,
+and lightly and predictably loaded, autoscaler is thus limited to the availability level of the complex Prometheus, which
+may be subject to load spikes and autoscaling evictions. By itself, this limitation is deemed acceptable, as the trigger
+thresholds, configured for PVC autoscaling, usually afford tolerance to substantial delays.
+
+Using the seed's cache Prometheus instance creates a dependency loop - `pvc-autoscaler` needs `prometheus-cache` in order
+to perform scaling, and `prometheus-cache` needs `pvc-autoscaler` to scale its database volume. That dependency results
+in one mode of persistent failure:
+1. `prometheus-cache` exhausts its storage before `pvc-autoscaler` resizes it
+2. `prometheus-cache` enters a fault loop, due to inability ot expand database
+3. `pvc-autoscaler` enters fault loop, due to driving metrics signal unavailable
+4. At this point, autoscaling seizes for all PVs on the seed
+
+To mitigate the risk of such failure, a larger initial storage request is used for `prometheus-cache` - 6Gi. It offers a
+1Gi headroom over the 5Gi storage retention limit, configured on `prometheus-cache`. Field data suggests 1Gi should
+suffice to accommodate the occasional usage spikes which exceed the data retention limit.
+
+#### Latency
+The pipeline which propagates the volume metrics, driving `pvc-autoscaler`, is as follows:
+1. `kubelet` collects PV state data and publishes resource metrics. Default poll interval: 1 minute
+   ([source](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/))
+2. `metrics-server` polls `kubelet` resource metrics. Default poll interval: 1 minute
+   ([source](https://github.com/kubernetes-sigs/metrics-server/blob/master/FAQ.md))
+3. `kube-apiserver` forwards metrics API calls to `metrics-server`, which is the registered APIService for metrics. Call
+   forwarding is instantaneous, so the delay is minimal.
+4. The seed's cache Prometheus polls the metrics provided by `kube-apiserver`. Default poll interval: 1 minute.
+5. `pvc-autoscaler` polls the seed's cache Prometheus instance. Poll interval: 1 minute.
+
+The reaction time, until `pvc-autoscaler` **initiates** a volume resize, can thus exceed 4 minutes. The resize operation
+itself, including the file system resize, can further take a few minutes. The cumulative delay, roughly estimated to up
+to 10 minutes, is mitigated via an appropriate safety margin, provided by the choice of utilisation level, configured
+as a scaling trigger threshold.
```

Comment:
How is the trigger threshold chosen? How is it configured?

Reply:
I am now mentioning the threshold and its value (90%) in the introductory paragraph. It is the `pvc-autoscaler`'s default. I reviewed it, and thought it reasonable. I did not come to it via some sort of optimisation calculation. I did however verify against Prometheus data, that it suffices to accommodate storage allocation spikes.

It can be configured via annotation on individual PVCs. However, I intend to leave it unspecified, and use the default.

Reply:
>it suffices to accommodate storage allocation spikes

As I read this, I realised that it's not quite true. I am aware of one category of spikes which can overwhelm the autoscaler with the proposed settings: Prometheus DB compaction causes limited, but particularly sharp surges in storage use (those are the spikes, subtly referred to in the _Limitations>Reliance on Prometheus_ section). For `cache-prometheus`, which is the critical case as it can result in persistent failure, I've mitigated those by setting a minimal volume size to 6Gi. Which results in scaling threshold of 10%=600Mi+ - enough to accommodate a spike.

I've been looking into an idea to enhance `pvc-autoscaler` with a behaviour which suits workloads with such sharp spikes, but I haven't reached a stage where I'd want to reflect it in the GEP.

Reply:
I'm in the process of adding support for an absolute `min-threshold` value (in bytes) to `pvc-autoscaler`. By setting an absolute minimum to the threshold, the aforementioned issue with sharp spikes when the volume is small, is completely resolved. I reflected this change by adding a new column to the "Initial volume size and MaxAllowed". For which, as an afterthought, a different title may now be in order.